Enhance the query for "Any" variations in filter by attribute widget

Konamiman opened the pull request:
### All Submissions:

* [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

_Actually implemented by @vedanshujain_

Enhance the query for variations with "Any" value in the filter by attribute widget.

The new query fixes a bug where variations were being counted twice: if a product was included in both the queries then it would be counted differently and added; e.g. when a product had two variations, one with "Any" attribute and other with a attribute that has a value.

The new query also optimizes performance, so that filter conditions can be improved and better indexes can be used.

### How to test the changes in this Pull Request:

See https://github.com/woocommerce/woocommerce/pull/27508.

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

Not needed since it's an improvement on https://github.com/woocommerce/woocommerce/pull/27508 which has the same milestone.

Konamiman committed:
Enhance the query for "Any" variations in filter by attribute widget.

The new query fixes a bug where variations were being counted twice:
if a product was included in both the queries then it would be counted
differently and added; e.g. when a product had two variations,
one with "Any" attribute and other with a attribute that has a value.

The new query also optimizes performance, so that filter conditions
can be improved and better indexes can be used.

jonathansadowski reviewed (dismissed):
Working well, good job.

Just noting for posterity:

> Set all the variations to "Out of stock". Reload the shop and verify that the value of the counters is now 1 (and now the product doesn't display in the shop). Set the variations to "In stock" again.

I still noticed a 2 for all counters when I don't have "Hide out of stock items from the catalog" enabled. If I enabled that in the settings, the counters became 1. I assume that this is by design.

vedanshujain reviewed (changes requested):
This still does not address the other bug where we are counting twice for the same variation. To reproduce on a fresh install:
1. Add a new attribute `Durability` on `wp-admin/edit.php?post_type=product&page=product_attributes` page.
2. After creating this attribute, add two values to this attribute, lets say `1year`, `2year`.
3. Create a new variable product. Add the above attribute to this product.
4. Create two variations for this product, one with attribute value `any` and other with `1year`. Set some price and publish the product.
5. Add the `Filter products by attribute` widget to the shop page. Set the filtering for attribute to `durability`.
6. Go to the shop page. You will see `2` counts for `1year` attribute value, whereas it should only be 1.

vedanshujain commented:
I am clearing the milestone because this is not ready for inclusion in 4.5 yet. But feel free to add it back if you think otherwise.

Konamiman committed:
Fix the SQL query to count products in the filter by attribute widget

The previous query was counting variable products twice when they
had a variation with a concrete value for the attribute and also
a variation with "Any..." value for the same attribute.

vedanshujain reviewed (approved):
LGTM!

jonathansadowski reviewed (approved):
Looks good with latest changes
